### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by karolukasik**
PeselNumber class

### CODE REVIEW

File: class-practice/src/main/java/com/karolina/Main.java [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package com.karolina;
+
+public class Main {
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/Main.java:3]: zbędne - do wyrzucenia

File: class-practice/pom.xml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>com.karolina</groupId>
+    <artifactId>class-practice</artifactId>
+    <version>1.0-SNAPSHOT</version>
+
+    <properties>
+        <maven.compiler.source>11</maven.compiler.source>
+        <maven.compiler.target>11</maven.compiler.target>
```
Comment (Jonarzz) [inline class-practice/pom.xml:13]: JDK11 ma 4,5 roku już - używaj bieżącej wersji (obecnie 20)

dzięki temu możesz np. skorzystać z `record`ów

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumber.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    public String extractSex() {
+        return SEX_DECODER.extract(pesel);
+    }
+
+
+    private boolean validatePeselCorrectness(String pesel) {
+        PeselNumberRule[] allRules = PeselNumberRule.values();
+        boolean areRulesFulfilled = true;
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumber.java:30]: do czego służy ta zmienna, skoro nie zmienia swojej wartości? nie różni się to od `return true`

z kolei skoro zawsze zwracane jest `true`, które i tak nie jest wykorzystywane, a w przypadku błędu walidacji leci wyjątek - po co zwracać cokolwiek?

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumber.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+        for (var rule : allRules) {
+            if (!rule.applyRule(pesel)) {
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumber.java:33]: `if (!rule.applyRule(pesel))` vs `if (rule.notValidFor(pesel))` / `if (rule.isNotMetBy(pesel))`

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumber.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        validatePeselCorrectness(trimmedPesel);
+        this.pesel = trimmedPesel;
+
+    }
+
+    public String extractDateOfBirth() {
+        return BIRTH_DATE_DECODER.extract(pesel);
+    }
+
+    public String extractSex() {
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumber.java:23]: jak masz to już na zewnątrz, to można byłoby przekazywać strategię wyciągania do metody, np. tak: `public String extractUsing(DataExtractor extractor)`, wtedy użycie wyglądałoby tak: `pesel.extractUsing(SEX_DECODER)` albo po zmianie nazw w enumie: `pesel.extract(SEX)`, `pesel.extract(BIRTH_DATE)`

obecnie przy dodawaniu kolejnej strategii, trzeba dodać też metodę tutaj

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselDecoder.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+            var date = LocalDate.of(birthYearNumber, birthMonthNumber, birthDayNumber);
+            var formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");
+            return date.format(formatter);
+        }
+
+    },
+    SEX_DECODER {
+        @Override
+        public String extract(String encodedData) {
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselDecoder.java:31]: doceniam możliwość zwrócenia dowolnej płci, ale jednak lepiej byłoby to zawrzeć w enumie, nawet jeśli byłoby tam dużo wartości, to klient API od razu wie, co dostanie w zwrotce

jaką wartosć ma `sex` w poniższym kodzie Twoim zdaniem, bez dodatkowego kontekstu? wielkie czy małe litery? w jakim języku? ;) 
`String sex = pesel.getSex()`

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselDecoder.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+            int controlDigit = getDigitAtIndexFrom(9, encodedData);
+            if (controlDigit % 2 == 0) {
+                return "woman";
+            } else {
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselDecoder.java:35]: `else` po returnie jest zbędny - można wypłaszczyć:

```java
if (...) {
    return 1;
}
return 0;
```

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselDecoder.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+            if (birthMonthNumber > 12) {
+                birthMonthNumber = birthMonthNumber - 20;
+                birthYearNumber = 2000 + birthYearNumber;
+            } else {
+                birthYearNumber = 1900 + birthYearNumber;
+            }
+
+            var date = LocalDate.of(birthYearNumber, birthMonthNumber, birthDayNumber);
+            var formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");
+            return date.format(formatter);
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselDecoder.java:25]: klient API musi się domyślać, jaki foramt dostanie albo dekompilować bibliotekę - jak wyciągana jest data, to powinien być zwrócony typ daty, ew. powinien być przyjmowany formatter i wtedy zwracany string

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselDecoder.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+            int birthDayNumber = generateDecimalNumberFromDigitsAtIndexes(4, 5, encodedData);
+
+            if (birthMonthNumber > 12) {
+                birthMonthNumber = birthMonthNumber - 20;
+                birthYearNumber = 2000 + birthYearNumber;
+            } else {
+                birthYearNumber = 1900 + birthYearNumber;
+            }
+
+            var date = LocalDate.of(birthYearNumber, birthMonthNumber, birthDayNumber);
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselDecoder.java:23]: https://www.youtube.com/watch?v=zsfEWLGgsEY

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/util/StringUtils.java [hunk 10]
```diff
@@ -0,0 +1,18 @@
+package com.karolina.tuesdaystasks.peselnumber.util;
+
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+public class StringUtils {
+
+    public static int getDigitAtIndexFrom(int index, String from) {
+        return Integer.parseInt(String.valueOf(from.charAt(index)));
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/util/StringUtils.java:9]: https://docs.oracle.com/javase/7/docs/api/java/lang/Character.html#getNumericValue(char)

w ramach ciekawostki polecam https://www.youtube.com/watch?v=QIEpZ0MGoBc

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/util/StringUtils.java [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+public class StringUtils {
+
+    public static int getDigitAtIndexFrom(int index, String from) {
+        return Integer.parseInt(String.valueOf(from.charAt(index)));
+    }
+
+    public static int generateDecimalNumberFromDigitsAtIndexes(int decimalIndex, int onesIndex, String from) {
+        int decimalNumber = getDigitAtIndexFrom(decimalIndex, from);
+        int onesNumber = getDigitAtIndexFrom(onesIndex, from);
+        return 10 * decimalNumber + onesNumber;
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/util/StringUtils.java:15]: czy mi się wydaje, czy ta skomplikowana logika to `value.substring(from, to + 1)` + `Integer.parseInt`? :p

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+        public String errorMessage() {
+            return "Incorrect length of input, should be 11 digits";
+        }
+
+
+    },
+    HAS_ONLY_DIGITS {
+        @Override
+        public boolean applyRule(String pesel) {
+            return pesel.matches("^[0-9]*$");
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java:22]: `^[0-9]{11}$` opędzluje 2 reguły jednocześnie

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+            return peselCorrectnessEquation(pesel);
+        }
+
+        @Override
+        public String errorMessage() {
+            return "Given number is not valid PESEL number";
+        }
+    };
+
+    private static boolean peselCorrectnessEquation(String pesel) {
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java:42]: `peselCorrectnessEquation` brzmi jak nazwa pola albo stałej, a nie metody

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+            sumOfDigits += digit;
+        }
+        if (sumOfDigits > 10) {
+            sumOfDigits = sumOfDigits % 10;
+        }
+        int controlNumber = getDigitAtIndexFrom(10, pesel);
+        return controlNumber == 10 - sumOfDigits;
+    }
+
+    private static int multiplyByWeightAndExtractOnesDigit(int i, int digit) {
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java:56]: i = index?

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+        @Override
+        public String errorMessage() {
+            return "Given number is not valid PESEL number";
+        }
+    };
+
+    private static boolean peselCorrectnessEquation(String pesel) {
+        int sumOfDigits = 0;
+        for (int i = 9; i >= 0; i--) {
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java:44]: ma to znaczenie, że nie jest od 0, tylko w drugą stronę?

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+            multiplicationResult = 9 * digit;
+        }
+        if (i == 6 || i == 2) {
+            multiplicationResult = 7 * digit;
+        }
+        if (i == 9 || i == 5 || i == 1) {
+            multiplicationResult = 3 * digit;
+        }
+
+        if (multiplicationResult > 10) {
```
Comment (Jonarzz) [inline class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java:68]: ten if-else to to samo co `return multiplicationResult % 10`

File: class-practice/src/test/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberTest.java [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+import org.junit.jupiter.api.function.Executable;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class PeselNumberTest {
+
+    @Test
+    public void constructorDoesNotAcceptNullParameter() {
+        Executable methodToExecute = () -> new PeselNumber(null);
+        Exception ex = assertThrows(IllegalArgumentException.class, methodToExecute);
```
Comment (Jonarzz) [inline class-practice/src/test/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberTest.java:15]: `assertThrows` lepiej pasuje do bloku "then" (analogicznie w pozostałych metodach)

File: class-practice/src/test/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberTest.java [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+    @ValueSource(strings = {"a2355896471", "aaaaaaaaaaa", "1234b567891"})
+    public void constructorDoesNotAcceptLettersInInputString(String value) {
+        Executable methodToExecute = () -> new PeselNumber(value);
+        Exception ex = assertThrows(IllegalArgumentException.class, methodToExecute);
+
+        assertEquals("Input should contain only numbers", ex.getMessage());
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"2", "123456789123"})
```
Comment (Jonarzz) [inline class-practice/src/test/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberTest.java:38]: warto byłoby dodać string o długości 10 jako edge case albo `@MethodSource` + `IntStream.concat` + `IntStream.rangeClosed` i sprawdzić wszystkie o długościach od 1 np. do 20 z wyłączeniem 11

File: class-practice/src/test/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberTest.java [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+        var pesel = new PeselNumber(value);
+
+        assertNotNull(pesel);
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"07321656897"})
+    public void applyRuleMethodReturnsTrueForStringOfLengthEleven(String value) {
+        var rule = PeselNumberRule.HAS_11_CHARS;
+        boolean applyingRuleResult = rule.applyRule(value);
```
Comment (Jonarzz) [inline class-practice/src/test/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberTest.java:75]: bloki given i when do rozdzielenia

File: class-practice/src/test/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberTest.java [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+    @ValueSource(strings = {" 07321656897 ", "\n00310453143 ", "\t87051456486 ", "\n88063039412\t"})
+    public void constructorCreateObjectWhenPeselIsValidWithWhitespaces(String value) {
+        var pesel = new PeselNumber(value);
+
+        assertNotNull(pesel);
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"07321656897"})
+    public void applyRuleMethodReturnsTrueForStringOfLengthEleven(String value) {
```
Comment (Jonarzz) [inline class-practice/src/test/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberTest.java:73]: te testy `PeselNumberRule` powinny być w oddzielnej klasie

**Review by Jonarzz — CHANGES_REQUESTED** [review]

File: class-practice/src/main/java/com/karolina/Main.java (hunk 1 above)
Comment (karolukasik) [reply class-practice/src/main/java/com/karolina/Main.java:3]: W sensie nie commitować tej klasy Main czy przenieść metodę main gdzieś do innej klasy?

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumber.java (hunk 5 above)
Comment (karolukasik) [reply class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumber.java:23]: Zrobione, ale niezgodne z wymogami zadania :P

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java (hunk 12 above)
Comment (karolukasik) [reply class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java:22]: I agree, ale czy warto łączyć dwie reguły w jedną? Zamysł był taki żeby jednak je rozdzielić

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java (hunk 15 above)
Comment (karolukasik) [reply class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java:44]: Nie ma znaczenia, miało kiedyś ale zmieniłam koncepcję :D Zmienione po bożemu

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java (hunk 14 above)
Comment (karolukasik) [reply class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java:56]: Si, zmienione

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java (hunk 16 above)
Comment (karolukasik) [reply class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java:68]: No nie, bo.... a jednak tak

File: class-practice/src/main/java/com/karolina/Main.java (hunk 1 above)
Comment (Jonarzz) [reply class-practice/src/main/java/com/karolina/Main.java:3]: nie commitować - kod weryfikujesz testami, a jest on de facto biblioteką klas do wykorzystania w ramach innego kodu, sam w sobie nic nie robi i nie ma sensu go "uruchamiać"

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumber.java (hunk 5 above)
Comment (Jonarzz) [reply class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumber.java:23]: zadanie nie przewidywało strategii, a jak już są, to warto z nich korzystać elastycznie :p

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java (hunk 12 above)
Comment (Jonarzz) [reply class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/PeselNumberRule.java:22]: a jak biznes wymyśli 3 wymagania:
- użytkownik musi być pełnoletni
- użytkownik musi być urodzony przed rokiem 2000
- użytkownik nie może mieć mniej niż 20 lat

to będą 3 ify? co najmniej 3 przypadki testowe - na pewno 

reguła, którą werfikuje mój regexp brzmi: "PESEL powinien mieć dokładnie 11 cyfr" (swoją drogą: `^\d{11}$`) - to zawiera w sobie obie i nawet biznes to zrozumie :p

File: class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/util/StringUtils.java (hunk 11 above)
Comment (karolukasik) [reply class-practice/src/main/java/com/karolina/tuesdaystasks/peselnumber/util/StringUtils.java:15]: Doobra nie popisuj się xD

**Review by Jonarzz — APPROVED** [review]
